scaler/scaler.go

```diff
@@ -106,51 +109,136 @@ func (s *Scaler) Run() error {
 		desired = current.MinSize
 	}
 
-	t := time.Now()
-
 	if desired > current.DesiredCount {
-		log.Printf("Scaling OUT 📈 to %d instances (currently %d)", desired, current.DesiredCount)
+		return s.scaleOut(desired, current)
+	}
 
-		err = s.autoscaling.SetDesiredCapacity(desired)
-		if err != nil {
-			return err
-		}
+	if current.DesiredCount > desired {
+		return s.scaleIn(desired, current)
+	}
 
-		log.Printf("↳ Set desired to %d (took %v)", desired, time.Now().Sub(t))
-	} else if current.DesiredCount > desired {
-		if s.scaleInParams.Disable {
-			log.Printf("Skipping scale IN, disabled")
-			return nil
-		}
+	log.Printf("No scaling required, currently %d", current.DesiredCount)
+	return nil
+}
+
+func (s *Scaler) scaleIn(desired int64, current AutoscaleGroupDetails) error {
+	if s.scaleInParams.Disable {
+		return nil
+	}
 
-		cooldownRemaining := s.scaleInParams.CooldownPeriod - time.Since(*s.scaleInParams.LastScaleInTime)
+	// If we've scaled down before, check if a cooldown should be enforced
+	if !s.scaleInParams.LastEvent.IsZero() {
+		cooldownRemaining := s.scaleInParams.CooldownPeriod - time.Since(s.scaleInParams.LastEvent)
 
 		if cooldownRemaining > 0 {
 			log.Printf("⏲ Want to scale IN but in cooldown for %d seconds", cooldownRemaining/time.Second)
 			return nil
 		}
+	}
 
-		minimumDesired := current.DesiredCount + s.scaleInParams.Adjustment
-		if desired < minimumDesired {
-			desired = minimumDesired
+	// Calculate the change in the desired count, will be negative
+	change := desired - current.DesiredCount
+
+	// Apply scaling factor if one is given
+	if factor := s.scaleInParams.Factor; factor != 0 {
+		// Use Floor to avoid never reaching upper bound
+		factoredChange := int64(math.Floor(float64(change) * factor))
+
+		if factoredChange < change {
+			log.Printf("👮‍️ Increasing scale-in of %d by factor of %0.2f",
```

Comment:
I'm not sure about the utility of having a scale-in factor >100% but if someone is crazy enough to configure that I guess we should honour it.

Reply:
Yeah, I thought the same 🤷🏼‍♂️